flash/data/data_source.py

```diff
@@ -0,0 +1,263 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import pathlib
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence, Tuple, Type, Union
+
+from pytorch_lightning.trainer.states import RunningStage
+from torchvision.datasets.folder import has_file_allowed_extension, make_dataset
+
+from flash.data.auto_dataset import AutoDataset, BaseAutoDataset, IterableAutoDataset
+from flash.data.data_pipeline import DataPipeline
+from flash.data.process import Preprocess, ProcessState
+from flash.data.utils import _STAGES_PREFIX, CurrentRunningStageFuncContext
+
+
+def has_len(data: Union[Sequence[Any], Iterable[Any]]) -> bool:
+    try:
+        len(data)
+        return True
+    except (TypeError, NotImplementedError):
+        return False
+
+
+@dataclass(unsafe_hash=True, frozen=True)
+class LabelsState(ProcessState):
+
+    labels: Optional[Sequence[str]]
+
+
+class MockDataset:
+
+    def __init__(self):
+        self.metadata = {}
+
+    def __setattr__(self, key, value):
+        if key != 'metadata':
+            self.metadata[key] = value
+        else:
+            object.__setattr__(self, key, value)
+
+
+class DataSource(ABC):
+
+    def __init__(
+        self,
+        train_data: Optional[Any] = None,
+        val_data: Optional[Any] = None,
+        test_data: Optional[Any] = None,
+        predict_data: Optional[Any] = None,
+    ):
+        super().__init__()
+
+        self.train_data = train_data
+        self.val_data = val_data
+        self.test_data = test_data
+        self.predict_data = predict_data
+
+        self._preprocess: Optional[Preprocess] = None
+
+    @property
+    def preprocess(self) -> Optional[Preprocess]:
+        return self._preprocess
+
+    @abstractmethod
+    def load_data(
+        self,
+        data: Any,
+        dataset: Optional[Any] = None
+    ) -> Iterable[Mapping[str, Any]]:  # TODO: decide what type this should be
```

Comment:
where does string suppose to point, specific files, images, ? It's a bit confusing what the `load_data` and `load_sample` methods are doing.

Reply:
Strings are the keys. So data can be any list of dicts like `{'input': ..., 'target': ...}`. So the `load_data` and `load_sample` are the same as before, just now have to be mappings rather than tuples, but I still need to add back the docstring :smiley: